### ISSUE TITLE
Add support for udb files

### ISSUE THREAD

**Issue by torognes** [torognes/vsearch#234]
See this thread on the VSEARCH Forum for details:

https://groups.google.com/forum/#!topic/vsearch-forum/3PEsOYRAaOQ

**Comment by torognes** [comment]
Support for UDB files added. The following commands are supported: `--makeudb_usearch`, `--udbinfo`, `--udbstats`, and `--udb2fasta`. The database specified by the `--db` option to the `--usearch_global` now detects UDB databases automatically.

The changes have not been tested extensively.

**Comment by torognes** [comment]
Several bugs with these new features have been discovered, so be careful. I'm working to fix them.

**Comment by colinbrislawn** [comment]
I'm curious to see how vsearch + indexing compares with other search methods that require indexing.

Here is how long sortmerna takes on a large database with a high identify reads:
```
time sortmerna --ref 99_otus.fasta,99_otus.idx --reads sub.fna \
--aligned sort.output --blast 1 -a 10

real	4m50.475s
user	24m54.180s
sys	0m42.957s
```

**Comment by GenomicaMicrob** [comment]
Hi, I tried to make it work with no avail, see:

> $ vsearch -v
> vsearch v2.4.4_linux_x86_64, 125.9GB RAM, 64 cores
> https://github.com/torognes/vsearch
> 
> $ vsearch --makeudb_usearch
> vsearch: unrecognized option '--makeudb_usearch'

I just downloaded 2.4.4 today and installed it, any ideas?

**Comment by torognes** [comment]
Hi, it is not in the latest release, only in the commits. You need to clone the repo and compile. It will be released when I cannot find any more bugs.

**Comment by colinbrislawn** [comment]
The current vsearch build works great. And it's fast!

```
time vsearch --usearch_global sub.fna --db 99_otus.udb --id .97 --uc test.uc \
--threads 10 --blast6out test.blast
vsearch v2.4.4_linux_x86_64, 62.9GB RAM, 24 cores
...
771599948 nt in 537233 seqs, min 900, max 3674, avg 1436
Searching 100%  
Matching query sequences: 5372 of 5372 (100.00%)

real	1m26.758s
user	8m31.125s
sys	0m6.000s
```

I think the bottleneck could be disk I/O; the `.udb` file is 3.8 GB total and when I'm reading it from the network, it's super slow. When loaded from local SSD, it's faster (`real	0m53.472s`).

**Comment by enormandeau** [comment]
Can it be confirmed that the makeudb_usearch command has been debugged in 2.5.0?

**Comment by torognes** [comment]
An inconsistency in results when using this feature has been reported (see the VSEARCH Forum), but I have not been able to reproduce the error yet. I will continue the effort to identify the problem.

Additional inconsistency reports that will help pinpointing the problem are highly appreciated.

Proceed with caution.

**Comment by kbseah** [comment]
Thanks for implementing the UDB option, it saves lots of time on re-indexing the same database when repeating runs. However I have another inconsistency to report, between the output from `usearch_global` with Fasta vs. UDB database. I'm using the v2.5.0 release compiled for Linux.

I'm running searches with vsearch usearch_global against the SILVA database, which have accession number and taxonomy string in the Fasta header, separated by a space. With the Fasta input, the output table includes the full header (with the taxonomy string). With the UDB file, only the part of the header before the space is reported in the output. It is a minor point (the accession number is already a unique identifier) but the output should be identical, and it breaks some downstream steps in an existing pipeline of mine. The hits are otherwise the same.

I hope that this report is useful, although it doesn't seem to be the same bug that's reported in the forum.

**Comment by torognes** [comment]
The source of the differences in results reported with and without udb databases might be due to the `--minseqlength` option. By default this option is 32 for the `--usearch_global` command and some other commands, but 1 for `--makeudb_usearch`. When some of the database sequences are shorter than 32 bp this could give different results unless the `--minseqlength 32` option is passed when running `--makeudb_usearch`. I will correct this problem.

**Comment by torognes** [comment]
It is important to include all the database-related options when running the `makeudb_usearch` command, including `--minseqlength`, `--maxseqlength`, `--dbmask`, `--hardmask`, `--notrunclabels`, and `--wordlength`.

**Comment by torognes** [comment]
The problem related to short sequences is fixed in release 2.5.1.

I am not aware of other UDB-related problems.
